api/v1/helper.go

```diff
@@ -258,55 +258,74 @@ func (p *SriovNetworkNodePolicy) Apply(state *SriovNetworkNodeState, merge bool)
 				Name:        iface.Name,
 				LinkType:    p.Spec.LinkType,
 				EswitchMode: p.Spec.EswitchMode,
+				NumVfs:      p.Spec.NumVfs,
 			}
-			var group *VfGroup
 			if p.Spec.NumVfs > 0 {
-				result.NumVfs = p.Spec.NumVfs
-				group, _ = p.generateVfGroup(&iface)
+				group, err := p.generateVfGroup(&iface)
+				if err != nil {
+					return err
+				}
+				result.VfGroups = []VfGroup{*group}
 				found := false
 				for i := range state.Spec.Interfaces {
 					if state.Spec.Interfaces[i].PciAddress == result.PciAddress {
 						found = true
-						// merge PF configurations when:
-						// 1. SR-IOV partition is configured
-						// 2. SR-IOV partition policies have the same priority
-						result = state.Spec.Interfaces[i].mergePfConfigs(result, merge)
-						result.VfGroups = state.Spec.Interfaces[i].mergeVfGroups(group)
+						state.Spec.Interfaces[i].mergeConfigs(&result, equalPriority)
 						state.Spec.Interfaces[i] = result
 						break
 					}
 				}
 				if !found {
-					result.VfGroups = []VfGroup{*group}
 					state.Spec.Interfaces = append(state.Spec.Interfaces, result)
 				}
 			}
 		}
 	}
+	return nil
 }
 
-func (iface Interface) mergePfConfigs(input Interface, merge bool) Interface {
-	if merge {
-		if input.Mtu < iface.Mtu {
-			input.Mtu = iface.Mtu
-		}
-		if input.NumVfs < iface.NumVfs {
-			input.NumVfs = iface.NumVfs
+// mergeConfigs merges configs from multiple polices where the last one has the
+// highest priority. This merge is dependent on: 1. SR-IOV partition is
+// configured with the #-notation in pfName and 2. SR-IOV policies have the same
+// priority.
+func (iface Interface) mergeConfigs(input *Interface, equalPriority bool) {
+	if !equalPriority {
```

Comment:
I think we would still want to allow VF partition policies with different priority numbers, for example:

Policy 1:
```
spec:
  deviceType: netdevice
  resourceName: kernel-device
  numVfs: 8
  priority: 50
  nicSelector:
    pfNames:
    - "ens1f0#0-3"
```

Policy 2:
```
spec:
  deviceType: vfio-pci
  resourceName: dpdk-device
  numVfs: 8
  priority: 70
  nicSelector:
    pfNames:
    - "ens1f0#4-7"
```

With current logic, policy-2 won't be applied if I understand correctly, but we should allow the policy-2 vfGroup be created since it doesn't conflict with policy-1 configuration.

This means the conditional check `!equalPriority` shall not be applied to vfGroup merging in line 308.

Reply:
From logical point of view I agree, these policies could be merged, but on the other hand wont that complicate the "priority's" field meaning? Now it is simple, merge only when same priority.
Why those policies cannot have same priority? If user explicitly defined the partitioning, the intention is to merge these, so in my opinion, they should have same priority.

Reply:
My concern is it may break existing deployment if we change current logic. For example, if user already has VF partition + different priorities settings, upgrade of operator will result in unexpected result.

I agree with you on that priority field may not necessary in most cases.

Reply:
Existing users can be instructed via documentation (or release notes), we are not changing any APIs. They would only have to adjust their priorities.
Additionally, this repo docs do not specify how exactly the priority field works, probably smth I can add to this PR.

Reply:
> Existing users can be instructed via documentation (or release notes), we are not changing any APIs. 
User may not want to change their configuration in a production environment that is working, to me this is like a regression I'd like to avoid.

> They would only have to adjust their priorities.
In my opinion, priority shall apply when there is conflict between policies (only apply high priority one over the other), but not when there is no conflict.

Reply:
Discussed this in Today's meeting.

In general, the `priority` field is not well documented (except of the one-line in the sriovnetworknodepolicy CRD)
more so, the interaction between VF partitioning and priority when merging configurations.

so we should definitely document this feature :)

AFAIU, today, if you use VF partitioning with multiple policies, if there is an overlap in VF group you might end up with an un-expected configuration (lower priority takes precedence)

I can think of additional edge cases, where you apply configuration on a per device basis and not per vf group like switchdev. e.g two sriovnetworkpolicies one with switchdev enabled the other without, same PF with different vf groups. (not sure if it would fail or not)

IMHO, if one wants to partition a PF's VFs to different resources he should do that in the same priority.

Reply:
I have updated the code to be backwards compatible. Added README changes.